Settings cache reader

leotianlizhan opened the pull request:
b/257074690
#no-changelog
## Goal
Add Crashlytics Settings to `FirebaseSessions`. 
## This PR
* Adds the "read local file" functionality of settings. 
* This also includes reading a so-called "cache-key" which provides meta-data to determine if current cache is expired. 
## Tests
Tests are split into 3 categories:
* Test settings API return correct values
  * default settings without cache
  * valid cache
* Test cache key expiration causes `isCacheExpired` to be marked as true
  * Can expire by TTL
  * Can expire by app version update
  * Can expire by GoogleID change
* Test corrupted cache file
  * Corrupted settings
  * Corrupted cache-key

samedson commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,167 @@
+    }
+    do {
+      objc_sync_enter(lock)
+      defer { objc_sync_exit(lock) }
```
This is smart - if I understand correctly you're using `do {...}` to create a block so that the `defer {objc_sync_exit...}` can evaluate at the end of the `do`, instead of at the end of the function.

I think you may be able to improve readability by making a lock function with a callback.

Eg.

```
func locked(callback: () -> ()) {
      objc_sync_enter(lock)
      defer { objc_sync_exit(lock) }
      callback()
}
```

Then use it with
```
locked() {
  /// do threading stuff
}
```

leotianlizhan replied:
That's a great idea, I love this syntactic sugar. Updated the PR to do basically this, but using an extension so the syntax now look like:
```
lock.synchronized {
...
}
```

samedson commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,167 @@
+      }
+    }
+    if appInfo.synthesizedVersion != cacheKey.appVersion {
+      Logger.logDebug("[Sessions:Settings] Settings expired because app version changed")
```
Kindof a nit - I believe we can leave off the `[Sessions:Settings]` and either have no prefix, or just use `[Settings]` as the prefix. Our logger by default adds a prefix to logs, eg:

```
 10.2.0 - [FirebaseSessions][I-SES000000] Logging events using fast QOS due to running on a simulator
```

leotianlizhan replied:
Done

samedson commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,167 @@
+  var synthesizedVersion: String { return "\(appDisplayVersion) (\(appBuildVersion))" }
+}
+
+private extension Data {
```
I believe it is customary to put extensions at the top of the file

leotianlizhan replied:
Done

samedson commented on FirebaseSessions/Sources/SettingsFileManager.swift:
```diff
@@ -0,0 +1,76 @@
+    do {
+      try fileManager.createDirectory(at: directoryUrl, withIntermediateDirectories: true)
+    } catch {
+      Logger.logDebug("SettingsFileManager: \(error)")
```
I might log this as a `logError` and add some context, eg.: `SettingsFileManager failed to create caches directory with error: \(error)`

leotianlizhan replied:
Done

samedson commented on FirebaseSessions/Sources/SettingsFileManager.swift:
```diff
@@ -0,0 +1,76 @@
+        try strongSelf.fileManager.removeItem(at: strongSelf.settingsCacheKeyPath)
+        try strongSelf.fileManager.removeItem(at: strongSelf.settingsCacheContentPath)
+      } catch {
+        Logger.logDebug("removeItem failed: \(error)")
```
Debug is good here, but similarly I might add more context to the message for our future selves.

leotianlizhan replied:
Done

ncooke3 commented on FirebaseSessions/Tests/Unit/SettingsTests.swift:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  func test_noCacheSaved_returnsDefaultSettings() {
+    assert(settings.isCacheExpired == true)
```
nit– I would favor XCTAssert for unit test assertions.

ncooke3 replied:
```suggestion
    XCTAssertTrue(settings.isCacheExpired)
```

leotianlizhan replied:
Done

ncooke3 reviewed (commented):
easy-to-read tests 👍

leotianlizhan commented on FirebaseSessions/Sources/SettingsFileManager.swift:
```diff
@@ -0,0 +1,76 @@
+  }
+}
+
+extension URL {
```
Done

leotianlizhan commented:
@ncooke3 @samedson Thanks for the review, addressed all the comments. However, I also added a `MockSettingsFileManager` to fix a test-flakiness issue with `removeCacheFilesAsync()`. Sometimes tests fail randomly because that method is running async in a dispatch queue, and it is fixed when I make it blocking/synchronous. However the way I did it feels kind of weird to me, is there a better way to mock this?

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+}
+
+private extension Data {
+  var dictionaryValue: [String: AnyObject]? {
```
What is the purpose of this dictionaryValue conversion?

leotianlizhan replied:
removed

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+  }
+}
+
+protocol SettingsProtocol {
```
Recommend adding some comments here for functions.

leotianlizhan replied:
Added some comments

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+  private static let cacheDurationSecondsDefault: TimeInterval = 60 * 60
+  private let fileManager: SettingsFileManagerProtocol
+  private let appInfo: ApplicationInfoProtocol
+  // Underscored variables may be accessed in different threads,
```
Is this a convention you are following?

leotianlizhan replied:
removed

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  var sessionsEnabled: Bool {
+    guard let enabled = settingsDictionary?["sessions_enabled"]?.boolValue else {
```
Lets define the Settings Flag Keys as Constants. Having them inline is hard to parse.

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+
+  var samplingRate: Double {
+    guard let rate = settingsDictionary?["sampling_rate"]?.doubleValue else {
+      return 1
```
return "1.0"

leotianlizhan replied:
Done

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+
+  var isCacheExpired: Bool {
+    guard settingsDictionary != nil else {
+      return true
```
Default is `true`? Does that mean we try to expire the cache every time the dictionary is empty?

When would the settingsDictionary be empty?

leotianlizhan replied:
outdated

visumickey commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+      removeCache()
+      return
+    }
+    guard cacheKey.googleAppID == googleAppID else {
```
Is this for the support for multi-apps? Does Crashlytics support multi-apps situation?

leotianlizhan replied:
It's for developers registering a new app id, not for multi-apps.

visumickey reviewed (commented):
I think we need a general design of this service. I'm confused with that we are trying to accomplish here. Is the general approach that we will have a separate fetching service, storing the configs and this PR is about taking the configs from the file to the memory? We need to align on the design here.

Secondly, most of the classes/functions/variables have zero to less comments. But can we ensure that we have enough comments to qualify the intent of the those variables/functions?

leotianlizhan commented on FirebaseSessions/Sources/SettingsCacheClient.swift:
```diff
@@ -0,0 +1,63 @@
+  init() {}
+
+  /// Converting to dictionary is O(1) because object conversion is O(1)
+  var cacheContent: [String: Any]? {
```
cacheKey is the meta-data associated with current cache, only used to determine if current cache is still valid (is it expired, did app version update, did developer register a new app id). It's like a "real life key" to the cache

visumickey replied:
Got it. Thanks for clarifying. Now it makes a lot more sense.

Can we call the variable `cachedContent` - CacheContent sounded like a verb - meaning the action of caching. We are referring to only the cachedContent here.

leotianlizhan commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,94 @@
+      cache.removeCache()
+      return true
+    }
+    if currentTime.timeIntervalSince(cacheKey.createdAt) > cacheDurationSeconds {
```
`cachekey.createdAt` is last fetched time. cache duration is part of the settings/cache content.

leotianlizhan commented on FirebaseSessions/Sources/SettingsCacheClient.swift:
```diff
@@ -0,0 +1,63 @@
+  /// UserDefaults holds values in memory, making access O(1) and synchronous within the app, while abstracting away async disk IO.
+  private let cache: UserDefaults = .standard
+
+  init() {}
```
Done

leotianlizhan commented on FirebaseSessions/Sources/SettingsCacheClient.swift:
```diff
@@ -0,0 +1,63 @@
+
+class SettingsCache: SettingsCacheClient {
+  private static let settingsVersion: Int = 1
+  private static let content: String = "settings"
```
done

leotianlizhan commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,173 @@
+
+import Foundation
+
+extension ApplicationInfoProtocol {
```
Added some comments

leotianlizhan commented on FirebaseSessions/Sources/Settings.swift:
```diff
@@ -0,0 +1,96 @@
+
+  var samplingRate: Double {
+    guard let rate = cache.cacheContent?["sampling_rate"] as? Double else {
+      return 1.0
```
Done
